### PR TITLE
Returning error messages from applicators

### PR DESCRIPTION
**Pull request by EmilsOzolins**
Work on https://github.com/harrel56/json-schema/issues/51 issue

### CODE REVIEW

**Comment by harrel56** [comment]
Funny stuff - GH doesn't make repo variables accessible in PRs from forks. I thought it's only the case for secrets :) Did a quick research how to work around this issue and it's not so straight forward - probably the easiest ad-hoc solution would be if you added required variables to your fork (see attached screenshot).

There still is a SonarCloud step which uses secrets - might disable it for fork PRs temporarily until I figure out how to do it correctly.
image

**Comment by harrel56** [comment]
BTW thanks for the willingness to contribute :) Let me know if you want to just take care of `oneOf` applicator or all of them. Because there probably will be some decisions to make:
- some applicators maybe don't need to produce any error message
- exact content of these new error messages
- I'm starting to feel that `Applicator` interface might be up for removal

**Comment by EmilsOzolins** [comment]
> Funny stuff - GH doesn't make repo variables accessible in PRs from forks. I thought it's only the case for secrets :) Did a quick research how to work around this issue and it's not so straight forward - probably the easiest ad-hoc solution would be if you added required variables to your fork (see attached screenshot).
> 
> There still is a SonarCloud step which uses secrets - might disable it for fork PRs temporarily until I figure out how to do it correctly. image

Thanks, I did exactly as you suggested. Let's see if it works now :)

**Comment by EmilsOzolins** [comment]
> BTW thanks for the willingness to contribute :) Let me know if you want to just take care of `oneOf` applicator or all of them. Because there probably will be some decisions to make:
> 
> * some applicators maybe don't need to produce any error message
> * exact content of these new error messages
> * I'm starting to feel that `Applicator` interface might be up for removal

I can take care of all of them but as I am not closely familiar with all of the specification, I only handled `anyOf`, `allOf`, `not` and `oneOf` for now. Also removed the Applicators interface as it seemed a bit redundant, as you already mentioned.

Let me know what you think about current proposal and if there is any other applicator I should take care of :)

File: src/main/java/dev/harrel/jsonschema/Applicators.java [hunk 1]
```diff
@@ -365,17 +410,28 @@ class OneOfEvaluator implements Applicator {
...
+        List<Integer> matchedIndexes = IntStream.range(0, refs.size())
+                .filter(i -> ctx.resolveInternalRefAndValidate(refs.get(i), node))
+                .boxed()
+                .collect(Collectors.toList());
+
+        if (matchedIndexes.size() == 1) {
+            return Result.success();
+        }
+
+        return Result.failure(String.format("Object matched against more than one schema. Matched schema indexes %s", matchedIndexes));
```
Comment (harrel56) [inline src/main/java/dev/harrel/jsonschema/Applicators.java:428]: Suggesting `Value matches against more than one schema. Matched schema indexes %s`

File: src/main/java/dev/harrel/jsonschema/Applicators.java [hunk 2]
```diff
@@ -365,17 +410,28 @@ class OneOfEvaluator implements Applicator {
...
+    }
+
+    @Override
+    public Result evaluate(EvaluationContext ctx, JsonNode node) {
+        List<Integer> matchedIndexes = IntStream.range(0, refs.size())
+                .filter(i -> ctx.resolveInternalRefAndValidate(refs.get(i), node))
+                .boxed()
+                .collect(Collectors.toList());
+
+        if (matchedIndexes.size() == 1) {
```
Comment (harrel56) [inline src/main/java/dev/harrel/jsonschema/Applicators.java:424]: Let's also handle case when there are no matched schemas, Suggesting `Value does not match against any of the schemas`

File: src/main/java/dev/harrel/jsonschema/Applicators.java [hunk 3]
```diff
@@ -384,12 +440,18 @@ class NotEvaluator implements Applicator {
...
-    public boolean apply(EvaluationContext ctx, JsonNode node) {
-        return !ctx.resolveInternalRefAndValidate(schemaUri, node);
+    public Set<String> getVocabularies() {
+        return APPLICATOR_VOCABULARY;
+    }
+
+    @Override
+    public Result evaluate(EvaluationContext ctx, JsonNode node) {
+        boolean valid = !ctx.resolveInternalRefAndValidate(schemaUri, node);
+        return valid ? Result.success() : Result.failure("Object matched against given schema but must not");
```
Comment (harrel56) [inline src/main/java/dev/harrel/jsonschema/Applicators.java:450]: Suggesting `Value matches against given schema but it must not`

File: src/main/java/dev/harrel/jsonschema/Applicators.java [hunk 4]
```diff
@@ -305,20 +322,29 @@ class IfThenElseEvaluator implements Applicator {
...
+
+    @Override
+    public Result evaluate(EvaluationContext ctx, JsonNode node) {
         if (ctx.resolveInternalRefAndValidate(ifRef, node)) {
-            return thenRef
+            boolean valid = thenRef
                     .map(ref -> ctx.resolveInternalRefAndValidate(ref, node))
                     .orElse(true);
+
+            return valid ? Result.success() : Result.failure("Object matched against If but failed matching against Then schema");
```
Comment (harrel56) [inline src/main/java/dev/harrel/jsonschema/Applicators.java:336]: Suggesting `Value matches against schema from 'if' but does not match against schema from 'then'`

File: src/main/java/dev/harrel/jsonschema/Applicators.java [hunk 5]
```diff
@@ -305,20 +322,29 @@ class IfThenElseEvaluator implements Applicator {
...
         } else {
-            return elseRef
+            boolean valid = elseRef
                     .map(ref -> ctx.resolveInternalRefAndValidate(ref, node))
                     .orElse(true);
+
+            return valid ? Result.success() : Result.failure("Object did not match against If but failed matching against Else schema");
```
Comment (harrel56) [inline src/main/java/dev/harrel/jsonschema/Applicators.java:342]: Suggesting `Value does not match against schema from 'if' and 'else'`

File: src/test/java/dev/harrel/jsonschema/ValidatorResultTest.java [hunk 6]
```diff
@@ -9,6 +9,104 @@
 import static org.assertj.core.api.Assertions.assertThat;
 
 class ValidatorResultTest {
+
+    @Test
+    void returnsErrorMessageWhenAnyOfFails() {
```
Comment (harrel56) [inline src/test/java/dev/harrel/jsonschema/ValidatorResultTest.java:14]: Could you also add test for `IfThenElse`?

**Review by harrel56 — COMMENTED** [review]
This looks really good - left a few comments, mainly error messages nitpicks. Still not 100% about the suggestions but I would at least try to be more consistent

**Comment by harrel56** [comment]
As for the failing workflows - I think I'll just test it out locally and merge. I'll try to fix it later on, so contribution would be smoother.

**Comment by harrel56** [comment]
Yeah, and I think the scope for now is fine. Gonna take care of whatever is left later on. Hopefully tomorrow all the changes could be released

File: src/main/java/dev/harrel/jsonschema/Applicators.java (hunk 2 above)
Comment (EmilsOzolins) [reply src/main/java/dev/harrel/jsonschema/Applicators.java:424]: Thanks, updated it to suggested error message

File: src/test/java/dev/harrel/jsonschema/ValidatorResultTest.java (hunk 6 above)
Comment (EmilsOzolins) [reply src/test/java/dev/harrel/jsonschema/ValidatorResultTest.java:67]: Thanks, missed that one but should be good now

**Comment by EmilsOzolins** [comment]
Thanks for the comments. A completely agree with your suggestions and have updated it accordingly

**Review by harrel56 — APPROVED** [review]
Looking good - appreciate your help :) gonna merge now

**Pull request merged**